python/torch_mlir/tools/import_onnx/__main__.py

```diff
@@ -48,9 +48,15 @@ def main(args):
         print(m.get_asm(assume_verified=not args.no_verify))
 
 
-def load_onnx_model(file_path: Path) -> onnx.ModelProto:
-    raw_model = onnx.load(file_path)
-    inferred_model = onnx.shape_inference.infer_shapes(raw_model)
+def load_onnx_model(file_path: Path, keep: bool) -> onnx.ModelProto:
+    # Do shape inference via files instead of in memory in order to handle
+    # models > 2 GB. See https://github.com/onnx/onnx/blob/main/docs/PythonAPIOverview.md#shape-inference-a-large-onnx-model-2gb
+    # for details about this technique.
+    inferred_path = file_path.with_stem(file_path.stem + '-inferred-shape')
```

Comment:
I think this is likely to interplay badly with automation and setups where the source directory is not read-write. We ultimately likely need to be a bit more switchy and have a flag for `--with-external-data` (as described here: https://github.com/onnx/onnx/blob/main/docs/PythonAPIOverview.md#loading-an-onnx-model-with-external-data)

Thinking of how to not boil the ocean. How about something like this:

* Use a `with tempfile.TemporaryDirectory(dir=flags.temp_dir, delete=not flags.keep_temps) as td:`
* `temp_path = Path(td.name, "inferred.onnx")`
* `onnx.shape_inference.infer_shapes_path(file_path, temp_path)`
* `inferred_model = onnx.load(temp_path, load_external_data=False)`
* `if flags.external_data: load_external_data_for_model(inferred_model, td.name)`

Or something like that. It can be made a lot better if anyone cares, but I'm not thrilled to do that without a reason. I think this will at least avoid some of the copies.

I hate proto based languages. Such a poor design decision.

Reply:
Oh, good idea!  I thought about this problem briefly but promptly forgot to deal with it.  On a previous product, we had the same issue with a temp index file.  We tried /tmp, user's home dir, and cwd, but had unhappy customers coming to us each time.  I think we finally gave them a temp_dir flag and the the wailing quieted down to a low grumbling.  :-D  I'll make these changes tomorrow when I'm feeling better.  Thanks for the feedback!

Reply:
Wait, I think I'm still unclear on what we're doing with external data and whether this is related to shape inference or just something else we should be supporting (for large models).  This statement in particular would seem to imply that we don't have a choice with external data location when doing shape inference:

> Current shape_inference supports models with external data, but for those models larger than 2GB, please use the model path for onnx.shape_inference.infer_shapes_path and the external data needs to be under the same directory.

Reply:
We don't really support it yet (I think) but I was trying to leave the door opened since we'll need to get it right soon.

There are a few modes for transferring back and forth and we'll probably need to elaborate them.

Reply:
It is implicated here because if you save the graph to a directory different from the data, you need to use the APIs explicitly to tell it where the original external data was.

Reply:
@stellaraccident New code with external data fails in MLIR verify.  Considering that the original code works, this is a regression.  :-(  Debugging it

Reply:
Ok, I think I'm ready for you again.  After tweaking the export size, now everything is working, small models and llama.  Added unit tests to test the command line on two small models in both internal and external data scenarios.

Unfortunately, I had to sacrifice running the checker on large models, as I couldn't find a way of pointing the file-based checker to the external data that llama does seem to generate.  In-memory checker fails in the same way as in-memory shape inference due to the 2 GB limit.

We could do away with the temp directory for the shape-inferred model and just put the inferred model wherever the user requests it.  If that location defaults to the same dir as the original model, then I would be able to run the checker.  Seems a bit beyond what's needed for now, but your call.